keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/DeliveryConfigRepository.kt

```diff
@@ -96,6 +103,30 @@ interface DeliveryConfigRepository : PeriodicallyCheckedRepository<DeliveryConfi
   ): List<ConstraintState>
 }
 
+/**
+ * Resolve [Locations] either from the spec or from [resource]'s [Environment]. One or the other
+ * must be non-null or an exception is thrown.
+ */
+fun <S : Locatable<SimpleLocations>> DeliveryConfigRepository.resolveLocations(resource: Resource<S>): SimpleLocations {
+  val specLocations = resource.spec.locations
+  val envLocations = environmentFor(resource.id).locations?.toSimpleLocations()
+  return checkNotNull(specLocations ?: envLocations) {
+    "Locations must be specified either in the resource spec, or its environment"
+  }
+}
+
+/**
+ * Resolve [Locations] either from the spec or from [resource]'s [Environment]. One or the other
+ * must be non-null or an exception is thrown.
+ */
+fun <S : Locatable<SubnetAwareLocations>> DeliveryConfigRepository.resolveLocations(resource: Resource<S>): SubnetAwareLocations {
+  val specLocations = resource.spec.locations
+  val envLocations = environmentFor(resource.id).locations
+  return checkNotNull(specLocations ?: envLocations) {
+    "Locations must be specified either in the resource spec, or its environment"
+  }
+}
```

Comment:
Standard extension methods used to resolve locations from `resource` or the `Environment` it belongs to.

I feel like I could probably do something clever with generic types and make this just one method.